src/tiled/stampbrush.cpp

```diff
@@ -380,13 +391,56 @@ void StampBrush::drawPreviewLayer(const QVector<QPoint> &list)
         QRegion paintedRegion;
         QVector<PaintOperation> operations;
         QHash<TileLayer *, QRegion> regionCache;
+        QHash<TileLayer *, TileLayer *> shiftedCopies;
 
         for (const QPoint &p : list) {
             const TileStampVariation variation = mStamp.randomVariation();
             mapDocument()->unifyTilesets(variation.map, mMissingTilesets);
 
             TileLayer *stamp = variation.tileLayer();
 
+            Map::StaggerAxis mapStaggerAxis = mapDocument()->map()->staggerAxis();
+            Map::StaggerIndex mapStaggerIndex = mapDocument()->map()->staggerIndex();
+            Map::StaggerIndex stampStaggerIndex = variation.map->staggerIndex();
+
+
+            //if staggered map, makes sure stamp stays the same
+            if (mapDocument()->map()->isStaggered()
+                    && ((mapStaggerAxis == Map::StaggerY)? stamp->height() > 1 : stamp->width() > 1)) {
+
+                if (mapStaggerAxis == Map::StaggerY) {
+                    bool topIsOdd = (p.y() - stamp->height() / 2) & 1;
+
+                    if ((stampStaggerIndex == mapStaggerIndex) == topIsOdd) {
+                        stamp = variation.tileLayer()->copy(variation.tileLayer()->bounds());
+                        shiftedCopies.insert(stamp, stamp);
```

Comment:
Before you go and make the copy and insert it, you should of course check whether it isn't already there (meaning as well, that you should use the original stamp as the key, and not the copy). If a shifted copy was already made for this stamp, just assign that to `stamp` and don't do the shifting again.